List all locks held by a transaction/query

thobe opened the pull request:
Adds a procedure called dbms.listActiveLocks(queryId) that lists all locks held by the transaction that executes the query with the given id.

Sample output from `CALL dbms.listQueries() YIELD queryId CALL dbms.listActiveLocks(queryId) YIELD mode, resourceType, resourceId RETURN *`:
```
+-----------------------------------------------------+
| mode        | queryId   | resourceId | resourceType |
+-----------------------------------------------------+
| "EXCLUSIVE" | "query-1" | 0          | "NODE"       |
| "EXCLUSIVE" | "query-1" | 1          | "NODE"       |
| "EXCLUSIVE" | "query-1" | 2          | "NODE"       |
| "EXCLUSIVE" | "query-1" | 3          | "NODE"       |
| "EXCLUSIVE" | "query-1" | 4          | "NODE"       |
| "SHARED"    | "query-1" | 0          | "SCHEMA"     |
| "SHARED"    | "query-3" | 0          | "SCHEMA"     |
+-----------------------------------------------------+
```

thobe committed:
List all locks held by a transaction

Adds a procedure called dbms.listActiveLocks(queryId) that lists all
locks held by the transaction that executes the query with the given id.

thobe commented:
Merging this should probably wait until #8579 has been merged, since it touches the same test files.

chrisvest commented:
Are they _query ids_ or _transaction ids_?

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -793,6 +793,12 @@ public void dispose()
+    {
+        StatementLocks locks = this.statementLocks;
+        return locks == null ? Collections.emptyList() : locks.activeLocks();
+    }
```
I think all the methods that can be accessed concurrently with the running transaction should have javadocs that clearly calls out this fact. The `markForTermination()` method is a good example.

thobe replied:
ok, I'll add that.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java:
```diff
@@ -126,4 +130,67 @@ void visit( ResourceType resourceType, long resourceId, String description, long
+        {
+            return "SHARED";
+        }
+    }
```
I guess you use this pattern to make the active locks collection use less heap memory?

thobe replied:
correct.

thobe replied:
... and also as a way to keep these string constants encapsulated.

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/locking/ActiveLocksListingCompatibility.java:
```diff
@@ -0,0 +1,60 @@
+                        new Locks.ActiveSharedLock( NODE, 3 ),
+                        new Locks.ActiveSharedLock( NODE, 4 ),
+                        new Locks.ActiveSharedLock( NODE, 5 ) ),
+                locks );
```
Do we really want to assert on the order? A lock manager implementation may be using hashed sets internally, and requiring an order would force a sort to happen.

thobe replied:
good point, thank you. I know I did unordered testing elsewhere. will fix.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/machines/locks/LeaderOnlyLockManager.java:
```diff
@@ -257,5 +258,11 @@ public int getLockSessionId()
+        public Collection<ActiveLock> activeLocks()
+        {
+            throw new UnsupportedOperationException( "not implemented" );
+        }
```
You should still be able to delegate this to the `localLocks`.

thobe replied:
ok.

chrisvest commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveLocksClient.java:
```diff
@@ -240,6 +241,12 @@ public int getLockSessionId()
+    @Override
+    public Collection<Locks.ActiveLock> activeLocks()
+    {
+        throw new UnsupportedOperationException( "not implemented" );
```
This one should also be able to use the `sharedLocks` and `exclusiveLocks` maps.

thobe replied:
ok.

chrisvest commented on enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/EnterpriseBuiltInDbmsProcedures.java:
```diff
@@ -245,6 +245,24 @@ public ProcedureResult( ProcedureSignature signature )
+        {
+            throwIfPresent( uncaught.getCauseIfOfType( InvalidArgumentsException.class ) );
+            throw uncaught;
+        }
```
Do we need `activeLocks()` to snapshot its state in a collection, or could it directly return a stream?

Why do we need to transform them to `ActiveLocksQueryResult` objects?

thobe replied:
I implemented it as snapshots since that was easier than streaming on the implementation level. Initially I had streams on the API level but then asked myself "who am I fooling", and changed the API to collections like it was in the implementation.

We need to convert to something with the appropriate public fields, not sure we want to have the internal API use such style of object, but maybe we do. Or perhaps we can turn this into some sort of visitor based (internal) API then have that directly construct the stream of the appropriate objects. That would be a more complicated API, but would imply less allocation.

thobe replied:
With it now changed to Steams a few places became simpler and could avoid allocation / copying.

chrisvest commented on enterprise/kernel/src/test/java/org/neo4j/kernel/enterprise/builtinprocs/ListQueriesProcedureTest.java:
```diff
@@ -71,36 +77,11 @@ public void shouldContainTheQueryItself() throws Exception
-        }, null, waitingWhileIn( GraphDatabaseFacade.class, "execute" ), 5, SECONDS );
-
-        try
+        String QUERY = "MATCH (n) SET n.v = n.v + 1";
```
It looks odd to me when local variables have all upper case names.

thobe replied:
I used uppercase as a highlight of the central aspect of the test, but I don't mind changing it.

chrisvest reviewed (commented):
@thobe I had a number of questions and comments that I'd like you to take a look at.

thobe commented:
Yes, these are Query IDs, not Transaction IDs, since we don't have any IDs fore live transactions. I think it would be better to introduce a Transaction ID, but that was deemed out of scope for the feature request that prompted this work.

chrisvest commented:
Okay. Would it make sense to name the procedure `listActiveLocksForQuery` so we have naming room for a future `listActiveLocks` that takes transaction ids, in case we want to add that later?

chrisvest commented:
Alternatively, if we add transaction ids later, I suppose the `dbms.listQueries()` could start returning those instead and we can keep the names and queries would keep working?

thobe commented:
if we add transaction ids later, `dbms.listQueries()` would return _both_ the transaction id and the query id. What we could do though is have the `listActiveLocks` procedure look at the input and do one thing if it is a transaction id, and another thing if it's a query id...

thobe committed:
Make Locks.Client.activeLocks() return a Stream

thobe committed:
Include lock mode in listQueries

If a query is waiting for a lock, include whether it is waiting for an
EXLUSIVE lock or a SHARED lock.

thobe committed:
Include a count of active locks in dbms.listQueries

chrisvest commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveLocksClient.java:
```diff
@@ -242,9 +247,21 @@ public int getLockSessionId()
+        return Stream.concat( // TODO: do we need to make these maps of locks ConcurrentHashMaps?
+                exclusiveLocks.entrySet().stream().flatMap( EXCLUSIVE_ACTIVE_LOCKS ),
+                sharedLocks.entrySet().stream().flatMap( SHARED_ACTIVE_LOCKS ) );
+    }
```
Ouch, that's a good point. HashMap is not safe for reads concurrent with writes, because a rehash will reverse the linked list collision buckets, which a concurrent reader than observe as a cycle in the chain, and thus end up in an infinite loop. `ConcurrentHashMap` has slower writes and uses more memory, but I think we probably need it for the lock maps. The maps that segregate by resource type could be turned into arrays and indexed by the resource types ordinal. The probability of seeing an old value in these arrays are probably negligible. It's unfortunate that there's no `AtomicReferenceArrayFieldUpdater`.

chrisvest reviewed (changes requested):
I had one comment about the use of `HashMap`s and thread safety in `SlaveLocksClient`. It also needs a rebase, which might also fix the failing test. Otherwise it looks good.

thobe commented:
@chrisvest yes, that was what I thought would be the right way to do it. I added those TODO notes so that someone would spot it in review and weigh in with another opinion on how to do that. Since we both have the same idea for how to structure it, I'll go ahead and make that change.

thobe commented:
@chrisvest the rebasing I was taking care of while you were writing your comment 😉

thobe committed:
Simplify SlaveLocksClient to avoid concurrency concerns

thobe commented:
@chrisvest fixed.

In a much simpler way. Since the local `Locks.Client` has all the lock information we need (except maybe re-entry counts), and is capable of supporting this API, we can just delegate to that and avoid any concurrency concern in `SlaveLocksClient`.

chrisvest commented:
Beautiful.
